rqd/rqd/rqcore.py

```diff
@@ -960,6 +960,93 @@ def __cleanup(self):
                 "Unable to close file: %s due to %s at %s",
                 self.runFrame.log_file, e, traceback.extract_tb(sys.exc_info()[2]))
 
+    def runSystemd(self):
+        """The steps required to handle a frame under linux using systemd-run"""
+
+        # pylint: disable = no-name-in-module, import-outside-toplevel
+        from cysystemd.reader import JournalReader, JournalOpenMode, Rule, JournalEvent
+
+        frameInfo = self.frameInfo
+        runFrame = self.runFrame
+
+        self.__createEnvVariables()
+        self.__writeHeader()
+
+        # TODO: Better handling if env vars are not set or not able to connect to the systemd-daemon
+        self.frameEnv['DBUS_SESSION_BUS_ADDRESS'] = os.environ['DBUS_SESSION_BUS_ADDRESS']
+        systemdUnitName = f"rqd-{frameInfo.frameId}.service"
+
+        def journalReaderThread():
+            rules = (
+                Rule("_SYSTEMD_USER_UNIT", systemdUnitName)
+            )
+            journalReader = JournalReader()
+            journalReader.open(JournalOpenMode.CURRENT_USER)
+            journalReader.seek_tail()
+            journalReader.add_filter(rules)
+            while True:
+                status = journalReader.wait(1)  # It will skip if there is no event for 1 second
+                if status == JournalEvent.NOP and frameInfo.forkedCommand.returncode is not None:
+                    break
+                for record in journalReader:
+                    self.rqlog.write(f"{record.data['MESSAGE']}\n",
+                                     prependTimestamp=rqd.rqconstants.RQD_PREPEND_TIMESTAMP)
```

Comment:
I wonder if a frame without logs would lead to a infinite loop here. Besides that, using a value set by another thread to control the flow of an open loop in a separate thread does't look safe without any loop bounds. I suggest adding a protection here for the number of attempts.

Another concern is the amount of threads that might be spawned when running on a large node running hundreds of frames. I had a similar implementation to stream the logs on rust/rqd and found out it was spawning too many threads on large hosts for log streaming.

Reply:
> Another concern is the amount of threads that might be spawned when running on a large node running hundreds of frames. I had a similar implementation to stream the logs on rust/rqd and found out it was spawning too many threads on large hosts for log streaming.

What kind of issues did you encounter?

My `/proc/sys/kernel/threads-max` is at `109006`, so it shouldn't be a problem. But I'll test with hundreds of dummy tasks locally that just spits out a lot of text.